### PR TITLE
Metrics instance policy addition to SDK

### PR DESCRIPTION
**Pull request by gspurthi**
Associated with the issue : kcpc/291

Changes included in the PR: Get and Set methods for metrics instance policy, update the set multiple instance policies method in a generalized way to accept any new policy addition in the future

### CODE REVIEW

File: instances.go [hunk 1]
```diff
@@ -108,6 +143,23 @@ func (c *Client) getInstancePolicy(ctx context.Context, policyType string, polic
 	return err
 }
 
+// GetMetricsInstancePolicy retrieves the metrics policy details associated with the instance
+// For more information can refer the Key Protect docs in the link below:
+// https://cloud.ibm.com/docs/key-protect?topic=key-protect-manage-metrics
```
Comment (robertmetcalf) [inline instances.go:148]: This URL does not exist. Coordinate with Kierra or Robert regarding documentation.

File: instances.go [hunk 2]
```diff
@@ -193,46 +278,59 @@ func (c *Client) SetAllowedNetworkInstancePolicy(ctx context.Context, enable boo
...
+		Policies: []InstancePolicy{policy},
+	}
+
+	err := c.setInstancePolicy(ctx, Metrics.String(), policyRequest)
+	if err != nil {
+		return err
 	}
 
-	if setAllowedNetwork {
+	return err
```
Comment (robertmetcalf) [inline instances.go:313]: You can just `return err` after calling `setInstancePolicy()`.

File: instances.go [hunk 3]
```diff
@@ -249,3 +347,36 @@ func (c *Client) SetInstancePolicies(ctx context.Context, setDualAuth, dualAuthE
...
 }
+
+type portsMetadata struct {
+	CollectionType string `json:"collectionType"`
+	NumberOfPorts  int    `json:"collectionTotal"`
+}
+
+type portResponse struct {
+	Metadata portsMetadata `json:"metadata"`
+	Ports    []privatePort `json:"resources"`
```
Comment (robertmetcalf) [inline instances.go:358]: This may be correct, but it looks out of place. Should the json be `ports` instead of `resources`?

File: instances.go [hunk 4]
```diff
@@ -249,3 +347,36 @@ func (c *Client) SetInstancePolicies(ctx context.Context, setDualAuth, dualAuthE
...
+}
+type privatePort struct {
+	PrivatePort int `json:"private_endpoint_port,omitempty"`
```
Comment (robertmetcalf) [inline instances.go:361]: The standard for our json is camel case. Is this used for encoding or decoding data? If encoding, it should be `private_endpoint_port` -> `privatePort` or `privateEndpointPort` (if all other port references are endpointPort).

File: instances.go [hunk 5]
```diff
@@ -249,3 +347,36 @@ func (c *Client) SetInstancePolicies(ctx context.Context, setDualAuth, dualAuthE
...
+		return 0, err
+	}
+
+	_, err = c.do(ctx, req, &portResponse)
+	if err != nil {
+		return 0, err
+	}
+
+	if len(portResponse.Ports) == 0 {
+		return 0, fmt.Errorf("No port number available. Please check the instance has an enabled allowedIP policy")
```
Comment (robertmetcalf) [inline instances.go:379]: Old: Please check the instance has an enabled allowedIP policy

New: Please check the instance and enable the allowedIP policy

File: instances.go (hunk 1 above)
Comment (gspurthi) [reply instances.go:148]: Currently it does not exist, Kierra provided me with this URL as the probable one

File: instances.go (hunk 3 above)
Comment (gspurthi) [reply instances.go:358]: resources is fine as we define our json output for all our entities as resources

**Comment by gspurthi** [comment]
@robertmetcalf The PR is against cli-integration branch and not master, I updated it. 
This PR is about metrics instance policy and not private port. 
Private port stuff is merged in 3Q.
I responded to your comments. Please verify your review.

File: instances.go (hunk 1 above)
Comment (robertmetcalf) [reply instances.go:148]: That's good for me.

File: instances.go [hunk 6]
```diff
@@ -252,46 +290,72 @@ func (c *Client) SetAllowedNetworkInstancePolicy(ctx context.Context, enable boo
...
+	return err
+}
+
+// SetInstancePolicies updates single or multiple policy details of an instance.
+func (c *Client) SetInstancePolicies(ctx context.Context, policies map[string]interface{}) error {
+	var resPolicies []InstancePolicy
+	var policyData PolicyData
+
+	for policyType, value := range policies {
+		if Contains(policyType) {
```
Comment (mrodden) [inline instances.go:334]:
```suggestion
		if isValidPolicyType(policyType) {
```

**Review by mrodden — CHANGES_REQUESTED** [review]
SetInstancePolicies(ctx, InstancePolicyOpts{
   DualAuthDeleteEnabled: true,
   AllowedNetworkEnabled: true,
   AllowedNetworks: "public-and-private",
   AllowedIPsEnabled: true,
   AllowedIPs: ["192.168.1.0/24", ...],
   MetricsEnabled: true,
});

SetInstancePolicies(ctx, InstancePolicyOpts{
   AllowedIPsEnabled: true,
   AllowedIPs: ["192.168.1.0/24", ...],
   MetricsEnabled: true,
});

// better for dealing with the "want to set these policies but do not want to change others"
// since the PolicyOpts attributes can be nil to reflect we aren't going to be modifying those
// policy types
SetInstancePolicies(ctx, InstancePolicyOpts{
   DualAuthDelete: DualAuthDelete{Enabled: true},
   AllowedNetwork: AllowedNetwork{Enabled: true, Networks: "public-and-private"},
   AllowedIP: AllowedIP{Enabled: true, IPs: ["192.168.1.0/24", ...]},
   Metrics: Metrics{Enabled: true}, 
});

SetInstancePolicies(ctx, InstancePolicyOpts{
   // DualAuthDelete: nil,
   AllowedNetwork: AllowedNetwork{Enabled: true, Networks: "public-and-private"},
   AllowedIP: AllowedIP{Enabled: true, IPs: ["192.168.1.0/24", ...]},
   Metrics: Metrics{Enabled: true}, 
});

File: instances.go [hunk 7]
```diff
@@ -16,22 +16,43 @@ package kp
...
 	"net/url"
 	"time"
 )
 
+type PolicyType int
+
 const (
 	//DualAuthDelete defines the policy type as dual auth delete
-	DualAuthDelete = "dualAuthDelete"
+	DualAuthDelete PolicyType = iota
```
Comment (robertmetcalf) [inline instances.go:30]: Why was this changed from strings to ints? There is a lot of casting to String() in this code. I feel that a more natural way would be leaving the consts as strings.

Then remove the type `PolicyType`.

Right after the `const` you could declare an array of strings to use in the `Contains()` function.

`var PolicyTypes = [...]string{AllowedIP, AllowedNetwork, DualAuthDelete, Metrics}`

File: instances.go [hunk 8]
```diff
@@ -16,22 +16,43 @@ package kp
 
 import (
 	"context"
+	"encoding/json"
```
Comment (robertmetcalf) [inline instances.go:19]: There is a typo on lines 290 and 296.

Old: dulaAuthPolicy

New: `dual`AuthPolicy

File: instances.go (hunk 7 above)
Comment (gspurthi) [reply instances.go:30]: This is an enum approach and we have used similar approach in KP code. 
However after discussion with Matt O, decided to revert it to strings in order not to disturb the global constants.

File: instances.go [hunk 9]
```diff
@@ -28,8 +28,11 @@ const (
 	//AllowedNetwork defines the policy type as allowed network
 	AllowedNetwork = "allowedNetwork"
 
-	//AllowedIP defines the polity type as allowed ip that are whitelisted
+	//AllowedIP defines the policy type as allowed ip that are whitelisted
 	AllowedIP = "allowedIP"
+
+	//Metrics defines the policy type as metrics
```
Comment (mrodden) [inline instances.go:34]:
```suggestion
	// Metrics defines the policy type as metrics
```

File: instances.go [hunk 10]
```diff
@@ -28,8 +28,11 @@ const (
 	//AllowedNetwork defines the policy type as allowed network
 	AllowedNetwork = "allowedNetwork"
 
-	//AllowedIP defines the polity type as allowed ip that are whitelisted
+	//AllowedIP defines the policy type as allowed ip that are whitelisted
```
Comment (mrodden) [inline instances.go:31]:
```suggestion
	// AllowedIP defines the policy type as allowed ip that are whitelisted
```

File: instances.go [hunk 11]
```diff
@@ -28,8 +28,11 @@ const (
 	//AllowedNetwork defines the policy type as allowed network
```
Comment (mrodden) [inline instances.go:28]:
```suggestion
	// AllowedNetwork defines the policy type as allowed network
```

**Review by mrodden — CHANGES_REQUESTED** [review]

**Review by mrodden — APPROVED** [review]
